tensorboard/plugins/pr_curve/summary.py

```diff
@@ -0,0 +1,208 @@
+    #   index + (num_thresholds * class_index)
+    class_indices = tf.reshape(tf.range(num_classes), (1, num_classes))
+    bucket_indices += num_thresholds * class_indices
+
+    with tf.name_scope('variables'):
+      # Now create the variables which correspond to the bucket values.
+      # These are flat arrays with num_thresholds value per class.
+      tp_buckets_v = tf.Variable(
+          tf.zeros([num_classes * num_thresholds], dtype=dtype),
+          name='tp_buckets',
+          trainable=False,
+          collections=[tf.GraphKeys.LOCAL_VARIABLES])
+      fp_buckets_v = tf.Variable(
+          tf.zeros([num_classes * num_thresholds], dtype=dtype),
+          name='fp_buckets',
+          trainable=False,
+          collections=[tf.GraphKeys.LOCAL_VARIABLES])
+
+    initialize_bucket_counts = tf.variables_initializer(
+        [tp_buckets_v, fp_buckets_v])
+    with tf.control_dependencies([initialize_bucket_counts]):
+      # Create the non-flat views with class_index as first dimension.
+      tp_buckets = tf.reshape(tp_buckets_v, (num_classes, -1))
+      fp_buckets = tf.reshape(fp_buckets_v, (num_classes, -1))
+
+      with tf.name_scope('update_op'):
+        # Use scatter_add to update the buckets.
+        update_tp = tf.scatter_add(
+            tp_buckets_v, bucket_indices, true_labels, use_locking=True)
+        update_fp = tf.scatter_add(
+            fp_buckets_v, bucket_indices, false_labels, use_locking=True)
+
+    with tf.control_dependencies([update_tp, update_fp]):
+      with tf.name_scope('metrics'):
+        thresholds = tf.cast(
+            tf.linspace(0.0, 1.0, num_thresholds), dtype=dtype)
+        # Set up the cumulative sums to compute the actual metrics.
+        tp_buckets = tf.cast(tp_buckets, tf.float32)
+        fp_buckets = tf.cast(fp_buckets, tf.float32)
+        tp = tf.cumsum(tp_buckets, reverse=True, axis=1, name='tp')
+        fp = tf.cumsum(fp_buckets, reverse=True, axis=1, name='fp')
+        # fn = sum(true_labels) - tp
+        #    = sum(tp_buckets) - tp
+        #    = tp[:, 0] - tp
+        # Similarly,
+        # tn = fp[:, 0] - fp
+        tn = tf.subtract(fp[:, 0:1], fp, name='tn')
+        fn = tf.subtract(tp[:, 0:1], tp, name='fn')
+
+        # Store the number of thresholds within the summary metadata because
+        # that value is constant for all pr curve summaries with the same tag.
+        summary_metadata = tf.SummaryMetadata()
+        pr_curve_plugin_data = pr_curve_pb2.PrCurvePluginData(
+            num_thresholds=num_thresholds)
+        summary_metadata.plugin_data.add(
+            plugin_name='pr_curve',
+            content=json_format.MessageToJson(pr_curve_plugin_data))
+
+        precision = tf.divide(
+            tf.maximum(_TINY_EPISILON, tp),
+            tf.maximum(_TINY_EPISILON, tf.add(tp, fp)))
+
+        # Use (1-fn/(tp+fn)) = tp/(tp+fn) so that at threshold, recall=1.
+        recall = tf.subtract(
```

Comment:
If these ops are unnamed, you might as well just write `recall = 1.0 - fn / tf.maximum(_TINY_EPSILON, tp + fn)` instead of using the functional notation. Likewise above: `precision = tf.maximum(_TINY_EPSILON, tp) / tf.maximum(_TINY_EPSILON, tp + fp)`.